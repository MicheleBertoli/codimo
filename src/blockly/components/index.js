/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */

function importAll(req) {
  req.keys().forEach(req);
}

// $FlowDoNotDisturb is a webpack function
importAll(require.context('./', true, /\.js$/));
