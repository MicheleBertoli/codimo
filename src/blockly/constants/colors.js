/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
export const ACTION = 345;
export const CONTAINER = 120;
