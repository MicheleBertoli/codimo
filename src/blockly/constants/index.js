/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
export * as colors from './colors';
export * as blockNames from './blockNames';
