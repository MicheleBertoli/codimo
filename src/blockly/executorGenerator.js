/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * This tester must evaluate every component.
 *
 * @flow
 */

// FIXME This function need to be more generic
// Right now only works for one specific function
function componentExecutionConfig(blockName: string) {
  // FIXME hardcoded 0
  Blockly.JavaScript[blockName] = () => (`0,${blockName};`);
}

export type ActorsToActions = Map<number, Array<string>>;
export type Executor = {|
  addBlockExecutor(blockName: string): void,
  parseInstructions(code: string): ActorsToActions,
|};
/**
 * This object is responsible for executing the code-as-string generated by blockly.
 * It allows to add every function used by each block.
 *
 * @return {Object} excecutor
 */
const executorGenerator = (): Executor => ({
  addBlockExecutor(blockName: string) {
    componentExecutionConfig(blockName);
  },
  /**
   * Each instruction follows this format:
   *      actor,action;
   * Where the semicolon is the separator.
   *
   * @param  {string} rawInstructions a sequence of raw instructions
   * @return {ActorsToActions}        a map with sequence of actions for each actor
   */
  parseInstructions: (rawInstructions: string): ActorsToActions => {
    const splittedRawInstructions = rawInstructions.split(';');
    const instructions = new Map();
    let actions;

    // Deletes the last & empty value
    splittedRawInstructions.pop();

    splittedRawInstructions.forEach((splittedRawInstruction) => {
      const elements = splittedRawInstruction.split(',');
      const key = parseInt(elements[0]);

      actions = instructions.get(key) || [];
      actions.push(elements[1]);

      instructions.set(key, actions);
    });

    return instructions;
  },
});

export default executorGenerator;
