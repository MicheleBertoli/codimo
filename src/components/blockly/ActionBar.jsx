/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
import React from 'react';
import styled from 'styled-components';

import Button from './Button';

const Bar = styled.div`
  align-items: center;
  background-color: #e3e3e3;
  border: 1px solid #ddd;
  border-bottom: 0;
  display: flex;
  height: 68px;
  width: 628px;
  & * {
    margin: 0 0 0 10px;
  }
`;

type Props = {|
  isExecuting: boolean,
  isStopped: boolean,
  handleStartGame(): void,
  handleResetGame(): void,
|};
// FIXME hardcoded title
const ActionBar = ({ isExecuting, isStopped, handleStartGame, handleResetGame }: Props) => (
  <Bar>
    {isStopped ?
      <Button
        title="▶ Dale play!"
        type="green"
        handleClick={handleStartGame}
      /> :
      <Button
        disabled={isExecuting}
        title="⏹ Reseteá!"
        type="orange"
        handleClick={handleResetGame}
      />
    }
  </Bar>
);

export default ActionBar;
