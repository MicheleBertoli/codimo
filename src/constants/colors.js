/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */

const WHITE = '#fff';

export const ACTION = '#b63668';
export const CONTAINER = '#369836';

export const BUTTON_COLORS = {
  blue: {
    bgColor: '#428bca',
    border: '#0a5a9c',
    color: WHITE,
  },
  green: {
    bgColor: '#5cb85c',
    border: '#1b7c1b',
    color: WHITE,
  },
  red: {
    bgColor: '#d9534f',
    border: '#be2f2b',
    color: WHITE,
  },
  orange: {
    bgColor: '#D3823F',
    border: '#B36320',
    color: WHITE,
  },
};
