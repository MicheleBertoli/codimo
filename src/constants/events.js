/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */

/////////////////
// Pointers
/////////////////

// To normalize touch and mouse click
export const POINTER_DOWN = 'pointerdown';
// Only click
export const CLICK = 'click';
// Only touch
export const TAP = 'tap';
