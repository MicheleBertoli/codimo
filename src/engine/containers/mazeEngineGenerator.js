/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
import { Container } from 'pixi.js';

import { ZERO, ONE } from 'constants/numbers';
import { type ActorsToActions } from 'blockly/executorGenerator';
import mazeGenerator, { type MazeData, type Maze } from 'engine/components/mazeGenerator';
import numberGenerator, { type NumberActor } from 'engine/components/numberGenerator';
import numericLineGenerator, {
  type NumericLineData,
  type NumericLine,
} from 'engine/components/numericLineGenerator';
import { MazePathError, MazeExitError, MazeWrongExitError } from 'engine/helpers/errors';
import { randomizeActorsConfig } from 'engine/helpers/randomConfigurations';

const numberHasLeftMazeConfig = (
  mazeData: MazeData,
  numericLineData: NumericLineData,
  numericLine: NumericLine,
) => (number: NumberActor, numberIndex: number): void => {
  const exitIdx = mazeData.exits.indexOf(number.position);

  if (exitIdx === -ONE) {
    throw new MazeExitError(numberIndex);
  }

  const exit = mazeData.exits[exitIdx];

  numericLine.receiveNumberAtPosition(number, numericLineData.accesses[exitIdx]);

  if (exit !== number.finalPosition) {
    throw new MazeWrongExitError(numberIndex);
  }
};
/* eslint-disable camelcase */
const directions = {
  move_forward: [ZERO, -ONE],
  move_right: [ONE, ZERO],
  move_backward: [ZERO, ONE],
  move_left: [-ONE, ZERO],
};
const directionsToWalls = {
  move_forward: 'top',
  move_right: 'right',
  move_backward: 'bottom',
  move_left: 'left',
};
/* eslint-enable */

const excecuteSetOfInstructionsConfig = (
  mazeData: MazeData,
  numbers: Array<NumberActor>,
  numberHasLeft: (number: NumberActor, numberIndex: number) => void,
) => async (instructions: ActorsToActions): Promise<void> => {
  const errors = [];

  for (const [numberPosition, actions] of instructions) {
    const number = numbers[numberPosition];

    for (let j = 0; j < actions.length; j++) {
      const direction = actions[j];
      const oldPosition = number.position;
      const newPosition = oldPosition.split(',')
                            .map((pos, i) => (parseInt(pos) + directions[direction][i]))
                            .join(',');
      const path = mazeData.path.get(oldPosition);

      if (!path || !path[directionsToWalls[direction]]) {
        errors.push(new MazePathError(numberPosition));
        break;
      }
      await number.updatePosition(newPosition);
    }
    if (errors[numberPosition]) {
      continue;
    }
    try {
      numberHasLeft(number, numberPosition);
    } catch (err) {
      errors.push(err);
    }
  }
  if (errors.length) {
    throw errors;
  }
};
const handleResetGameConfig = (
  randomizeActors: () => Array<number>,
  numbers: Array<NumberActor>,
  maze: Maze,
  actorsPositions: Array<[number, number]>,
) => () => {
  const newActors = randomizeActors();

  numbers.forEach((number, idx) => {
    number.changeActor(newActors[actorsPositions[idx][1]]);
    number.view.setParent(maze.view);
    number.resetPosition();
  });
};

export type GameDifficulty = 'easy' | 'normal' | 'hard';
export type Engine = {|
  view: Container,
  excecuteSetOfInstructions(instructions: ActorsToActions): Promise<void>,
  handleResetGame(): void,
|};
export default function mazeEngineGenerator(
  mazeData: MazeData,
  numericLineData: NumericLineData,
  difficulty: GameDifficulty,
): Engine {
  const view = new Container();
  const numericLine = numericLineGenerator(
    numericLineData.statics,
    mazeData.size,
    mazeData.margin,
  );

  const maze = mazeGenerator(mazeData);
  const randomizeActors = randomizeActorsConfig(
    numericLineData.statics,
    numericLineData.accesses,
    difficulty,
  );
  const randomActors = randomizeActors();
  const numbers: Array<NumberActor> = mazeData.actorsPositions.map((actorPositions) => {
    // El actor también tiene que recibir en qué posición sale correctamente, eso se tiene que chequear!!
    // El número SIEMPRE tiene que entrar a la recta, por más que sea erróneo su lugar
    const actor = numberGenerator(
      randomActors[actorPositions[1]],
      mazeData.accesses[actorPositions[0]],
      mazeData.exits[actorPositions[1]],
      mazeData.size,
      mazeData.margin,
    );

    maze.view.addChild(actor.view);

    return actor;
  });

  maze.view.y = numericLine.view.height - mazeData.margin;

  view.addChild(maze.view, numericLine.view);

  return {
    view,
    excecuteSetOfInstructions: excecuteSetOfInstructionsConfig(
      mazeData, numbers, numberHasLeftMazeConfig(mazeData, numericLineData, numericLine),
    ),
    handleResetGame: handleResetGameConfig(randomizeActors, numbers, maze, mazeData.actorsPositions),
  };
}
