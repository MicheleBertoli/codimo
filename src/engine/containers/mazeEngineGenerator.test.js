/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
/* eslint-disable no-magic-numbers */
import gameMetadataDataWithoutBlocks from 'test/gameMetadataDataWithoutBlocks.json';
import { blockNames } from 'blockly/constants';

import mazeEngineGenerator from './mazeEngineGenerator';

const ACTOR = 0;
const ONE = 1;
const TWO = 2;
const { mazeData, numericLineData } = gameMetadataDataWithoutBlocks;

mazeData.path = new Map(mazeData.path);

describe('engine > containers > mazeEngineGenerator', () => {
  it('should parse an array of correct actions and return a response', async () => {
    const mazeEngine = mazeEngineGenerator(mazeData, numericLineData);
    const actions = new Map();

    actions.set(ACTOR, [
      blockNames.MOVE_FORWARD,
      blockNames.MOVE_FORWARD,
    ]);

    try {
      await mazeEngine.excecuteSetOfInstructions(actions);
    } catch (errors) {
      const error = errors[0];

      expect(errors.length).toBe(ONE);
      expect(error.actor).toBe(ACTOR);
      expect(error.name).toBe('MazeExitError');
      expect(error.message).toMatchSnapshot();
    }
  });
  it('should throw a path Error for ACTOR 1 but exit error for ACTOR 2', async () => {
    const ACTOR2 = ONE;
    const newMazeData = {
      ...mazeData,
      accesses: [mazeData.accesses[0], mazeData.accesses[0]],
      exits: [mazeData.exits[0], mazeData.exits[0]],
    };
    const newNumericLineData = {
      statics: [1, null, 5, null, 9],
      accesses: [1, 3],
    };
    const mazeEngine = mazeEngineGenerator(newMazeData, newNumericLineData);
    const actions = new Map();

    actions.set(ACTOR, [
      blockNames.MOVE_LEFT,
    ]);
    actions.set(ACTOR2, [
      blockNames.MOVE_FORWARD,
      blockNames.MOVE_FORWARD,
    ]);

    try {
      await mazeEngine.excecuteSetOfInstructions(actions);
    } catch (errors) {
      expect(errors).toBeInstanceOf(Array);
      expect(errors.length).toBe(TWO);
      expect(errors[0].name).toBe('MazePathError');
      expect(errors[1].name).toBe('MazeExitError');
    }
  });
});
