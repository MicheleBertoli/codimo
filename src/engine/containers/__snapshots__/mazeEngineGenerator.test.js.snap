// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`engine > containers > mazeEngineGenerator should parse an array of correct \`instructions\` and return a response 1`] = `undefined`;
